### PR TITLE
API documentation improvements

### PR DESCRIPTION
**Pull request by sipa**

### CODE REVIEW

**Comment by dcousens** [comment]
Reviewed with [w=0?](https://github.com/bitcoin/secp256k1/pull/289/files?w=0).

utACK

File: include/secp256k1.h [hunk 1]
```diff
@@ -361,60 +419,113 @@ SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_seckey_verify(
...
 ) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3) SECP256K1_ARG_NONNULL(4);
 
-/** Import a private key in DER format. */
+/** Import a private key in DER format.
+ * Returns: 1 if a private key was extracted.
+ * In:  ctx:        pointer to a context object (cannot be NULL).
+ *      privkey:    pointer to a private key in DER format (cannot be NULL).
+ *      privkeylen: length of the DER private key pointed to be privkey.
+ * Out: seckey:     pointer to a 32-byte array for storing the private key.
+ *                  (cannot be NULL).
```
Comment (dcousens) [inline include/secp256k1.h]: Any reason this functions in/out parameters are not in the same order as the documentation?

File: include/secp256k1.h [hunk 2]
```diff
@@ -336,24 +392,26 @@ int secp256k1_ecdsa_sign(
  *  Out:     pubkey:     pointer to the recoved public key (cannot be NULL)
  */
 SECP256K1_WARN_UNUSED_RESULT int secp256k1_ecdsa_recover(
-  const secp256k1_context_t* ctx,
-  const unsigned char *msg32,
-  const secp256k1_ecdsa_signature_t *sig,
-  secp256k1_pubkey_t *pubkey
+    const secp256k1_context_t* ctx,
+    const unsigned char *msg32,
+    const secp256k1_ecdsa_signature_t *sig,
```
Comment (dcousens) [inline include/secp256k1.h]: Documentation calls this `sig64`,  its `sig` here.

File: src/bench_internal.c [hunk 3]
```diff
@@ -282,6 +283,22 @@ void bench_rfc6979_hmac_sha256(void* arg) {
...
+    }
+}
+
+void bench_context_sign(void* arg) {
+    (void)arg;
+    int i;
+    for (i = 0; i < 200; i++) {
+        secp256k1_context_destroy(secp256k1_context_create(SECP256K1_CONTEXT_SIGN));
+    }
+}
```
Comment (dcousens) [inline src/bench_internal.c]: For posterity, it should probably be mentioned this PR also adds 2 new benchmarks.

File: include/secp256k1.h [hunk 4]
```diff
@@ -5,6 +5,83 @@
...
 
+/** Opaque data structure that holds context information (precomputed tables etc.).
+ *
+ *  The purpose of context structures is to cache large precomputed data tables
+ *  that are expensive to construct, and also to maintain the randomization data
+ *  for blinding.
+ *
+ *  Do not create a new context object for each operation, as construction is
+ *  far slower than all other API calls (~100 times slower than an ECDSA
+ *  verification), and then use the context for all other operations.
```
Comment (apoelstra) [inline include/secp256k1.h]: I think part of this sentence is missing; you say "do not do this .. and then" rather than "do not do this .. instead do this, and then".

File: include/secp256k1.h [hunk 5]
```diff
@@ -5,6 +5,83 @@
...
+ *
+ *  A constructed context can safely be used from multiple threads
+ *  simultaneously, but some API calls need exclusive access, in particular
+ *  secp256k1_context_destroy and secp256k1_context_randomize.
```
Comment (apoelstra) [inline include/secp256k1.h]: Do we want to drop the comment that `const` makes the difference between requiring exclusive access and not?

File: include/secp256k1.h (hunk 1 above)
Comment (sipa) [inline include/secp256k1.h]: I'd to have a convention of having all functions AND all documentation to have order: ctx/out/inout/in or ctx/in/inout/out.

**Comment by sipa** [comment]
Addressed nits.

File: src/bench_internal.c (hunk 3 above)
Comment (sipa) [inline src/bench_internal.c]: Done, split into a separate commit.

File: include/secp256k1.h (hunk 1 above)
Comment (dcousens) [inline include/secp256k1.h]: I'm personally in favour of ctx/in/inout/out

File: include/secp256k1.h (hunk 2 above)
Comment (dcousens) [inline include/secp256k1.h]: ping @sipa,  missed nit or ?

File: include/secp256k1.h [hunk 6]
```diff
@@ -336,24 +393,26 @@ int secp256k1_ecdsa_sign(
  *  Out:     pubkey:     pointer to the recoved public key (cannot be NULL)
```
Comment (dcousens) [inline include/secp256k1.h]: ^`sig64`

File: include/secp256k1.h (hunk 2 above)
Comment (sipa) [inline include/secp256k1.h]: Fixed.

File: include/secp256k1.h (hunk 1 above)
Comment (sipa) [inline include/secp256k1.h]: There is a convention within C APIs to use out/in order, to mimic assignment.

Comment (dcousens) [inline include/secp256k1.h]: I guess that makes sense,  but,  I wouldn't be able to change the habit of reading things in terms of INPUT / OUTPUT (IO).

It'd be similar to seeing an anonymous function callback being used with the callback passed before arguments to the function.

Comment (apoelstra) [inline include/secp256k1.h]: I prefer out/in order, it is extremely common in C API's.

File: src/bench_internal.c [hunk 7]
```diff
@@ -282,6 +283,22 @@ void bench_rfc6979_hmac_sha256(void* arg) {
...
+    (void)arg;
+    int i;
+    for (i = 0; i < 20; i++) {
+        secp256k1_context_destroy(secp256k1_context_create(SECP256K1_CONTEXT_VERIFY));
+    }
+}
+
+void bench_context_sign(void* arg) {
+    (void)arg;
+    int i;
```
Comment (apoelstra) [inline src/bench_internal.c]: I'm getting a "mixed declarations and code" warning for having `int i` after the `(void)arg`. Ditto in the above function.

Comment (sipa) [inline src/bench_internal.c]: Fixed.

**Comment by apoelstra** [comment]
ACK

**Comment by gmaxwell** [comment]
ACK.

File: include/secp256k1.h [hunk 8]
```diff
@@ -5,6 +5,84 @@
...
+} secp256k1_pubkey_t;
+
+/** Opaque data structured that holds a parsed ECDSA signature, optionally
+ *  supporting pubkey recovery.
+ *
+ *  The exact representation of data inside is implementation defined and not
+ *  guaranteed to be portable between different platforms or versions. It is
+ *  however guaranteed to be 65 bytes in size, and can be safely copied/moved.
+ *  If you need to convert to a format suitable for storage or transmission, use
+ *  the secp256k1_ecdsa_signature_serialize_* and
```
Comment (droark) [inline include/secp256k1.h]: After-the-fact nit: "secp256k1_ecdsa_signature_serialize_*" appears twice. I'm guessing this is a mistake.
